subprojects/core/src/main/java/org/gradle/internal/build/PlannedNodeGraph.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.build;
+
+import com.google.common.collect.ImmutableList;
+import org.gradle.execution.plan.Node;
+import org.gradle.execution.plan.PlannedNodeInternal;
+import org.gradle.execution.plan.ToPlannedNodeConverter;
+import org.gradle.execution.plan.ToPlannedNodeConverterRegistry;
+import org.gradle.internal.taskgraph.CalculateTaskGraphBuildOperationType.PlannedNode;
+import org.gradle.internal.taskgraph.CalculateTaskGraphBuildOperationType.TaskIdentity;
+import org.gradle.internal.taskgraph.NodeIdentity;
+import org.gradle.internal.taskgraph.NodeIdentity.NodeType;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import static java.util.Collections.newSetFromMap;
+
+/**
+ * A graph of planned nodes that supports extracting sub-graphs of different {@link DetailLevel levels of detail}
+ * depending on what node types are supported by the consumer.
+ *
+ * @see Collector
+ */
+public class PlannedNodeGraph {
+
+    private final DetailLevel detailLevel;
+    private final List<PlannedNodeInternal> plannedNodes;
+
+    public PlannedNodeGraph(DetailLevel detailLevel, List<PlannedNodeInternal> plannedNodes) {
+        this.detailLevel = detailLevel;
+        this.plannedNodes = ImmutableList.copyOf(plannedNodes);
+    }
+
+    /**
+     * Returns the nodes from a subgraph that are of the given or lower detail level
+     * (lower if the requested detail level is higher than the available detail level).
+     */
+    public List<? extends PlannedNode> getNodes(DetailLevel detailLevel) {
+        if (detailLevel.level >= this.detailLevel.level) {
+            return plannedNodes;
+        }
+
+        return computePlan(detailLevel);
+    }
+
+    private List<PlannedNodeInternal> computePlan(DetailLevel detailLevel) {
+        HashMap<NodeIdentity, List<? extends NodeIdentity>> plannedNodeByIdentity = new HashMap<>();
```

Comment:
```suggestion
        HashMap<NodeIdentity, List<? extends NodeIdentity>> plannedNodeDependenciesByIdentity = new HashMap<>();
```

These are the dependencies per node identity, not the nodes, right?

Reply:
Fixed